src/plugins/spec/spec-new.c

```diff
@@ -0,0 +1,762 @@
+		if (elektraStrCmp (strchr (keyName (current), '/'), name) == 0)
+		{
+			return current;
+		}
+	}
+	return 0;
+}
+
+/**
+ * Get the key which contains the array size as meta key.
+ *
+ * @param specKeys the specification keys
+ * @param specKey the specification key to use for lookup
+ * @return the key which contains the array size
+ * @retval 0 - if the key was not found
+ */
+static Key * getArraySizeOfArrayParent (KeySet * specKeys, Key * specKey)
+{
+	const char * specKeyName = keyName (specKey);
+	char * copiedKeyName = elektraMalloc (elektraStrLen (specKeyName) + 1);
+	strcpy (copiedKeyName, (char *) specKeyName);
+
+	char * rest = NULL;
+	char * arrayParent = strtok_r (copiedKeyName, "#", &rest);
+	arrayParent[strlen (arrayParent) - 1] = '\0';
+
+	Key * key = getMatchingKeyFromKeySet (specKeys, strchr (arrayParent, '/'));
+
+	elektraFree (copiedKeyName);
+
+	return key;
+}
+
+/**
+ * Get the number of array characters in an array specification key name.
+ *
+ * @param specKey the specification key to use
+ * @return number of array characters in the array specification key name
+ */
+static int getNumberOfArrayCharactersInSpecName (Key * specKey)
+{
+	char * withoutNamespace = strchr (keyName (specKey), '/');
+
+	int count = 0;
+
+	for (size_t i = 0; i < elektraStrLen (withoutNamespace); i++)
+	{
+		if (withoutNamespace[i] == '#')
+		{
+			count++;
+		}
+	}
+
+	return count;
+}
+
+/**
+ * The positions of where the `#` occurs in the specification key.
+ *
+ * @param keyNameWithoutNamespace the key name without the namespace
+ * @param arrayPositions the array to hold all the positions of the `#` in the specification key
+ * @param arraySize number of `#` in the whole specification key
+ */
+static void setArrayPositions (const char * keyNameWithoutNamespace, int * arrayPositions)
+{
+	int arrPos = 0;
+	for (int i = 0; i < (int) elektraStrLen (keyNameWithoutNamespace); i++)
+	{
+		if (keyNameWithoutNamespace[i] == '#')
+		{
+			arrayPositions[arrPos++] = i;
+		}
+	}
+}
+
+/**
+ * Validate the array size. If it does not find any array size a warning is appended to the parent key.
+ *
+ * @param ks the key set to look for the key
+ * @param specKeys the specification keys
+ * @param specKey the array specification key
+ * @retval true - if the array size is valid
+ * @retval false - if the array size was not found
+ * @retval false - if the array size is not valid
+ */
+static bool isValidArraySize (KeySet * ks, KeySet * specKeys, Key * specKey)
+{
+	Key * key = ksLookupByName (ks, strchr (keyName (specKey), '/'), 0);
+	Key * keyToFetchArraySizeFrom = key == NULL ? getArraySizeOfArrayParent (specKeys, specKey) : key;
+	const Key * arrayMetaKey = key == NULL ? keyGetMeta (keyToFetchArraySizeFrom, "array") : keyGetMeta (key, "array");
+
+	if (arrayMetaKey == 0)
+	{
+		// no array size found, skip
+		return true;
+	}
+
+	if (!validateArraySize (keyToFetchArraySizeFrom, specKey))
+	{
+		return false;
+	}
+
+	return true;
```

Comment:
Could be simplified.

```suggestion
	return validateArraySize (keyToFetchArraySizeFrom, specKey);
```